### PR TITLE
feat(perception_utils): add a function that gets the highest confide…

### PR DESCRIPTION
**Pull request by purewater0901**
Signed-off-by: yutaka <purewater0901@gmail.com>

## Description
Add a function that extracts a predicted path that has the highest confidence from predicted paths

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/perception_utils/test/src/test_predicted_path_utils.cpp [hunk 1]
```diff
@@ -60,6 +60,56 @@ PredictedPath createTestPredictedPath(
...
+    std::vector<PredictedPath> predicted_paths(10);
+    for (size_t i = 0; i < 10; ++i) {
+      const auto path = createTestPredictedPath((i + 1) * 10, 0.1, (i + 1) * 1.0, 0.0, 0.0, 0.1);
+      predicted_paths.at(i) = path;
+    }
+    const auto path = getPredictedPathWithHighestConfidence(predicted_paths);
+
+    EXPECT_NEAR(path.confidence, 0.1, epsilon);
+    EXPECT_NEAR(path.path.size(), 10, epsilon);
+  }
```
Comment (yukkysaito) [inline common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: If it is the same confidence, which one should be selected?
cc @miursh

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @yukkysaito Currently, this function selects the first predicted path in the predicted paths. Do you have any other ideas when all paths have same confidence?

Comment (yukkysaito) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @purewater0901 @miursh 
umm...return multiple path. 
how about you?

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: The function aims to return a single path, so I personally don't want to return multiple paths.

Comment (yukkysaito) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: Yes. I know.
But which one should be selected?
I think it is not correct that this function selects the first predicted path in the predicted paths.
@kenji-miyake @xmfcx @mitsudome-r any idea? :pray:

Comment (kenji-miyake) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: I feel it depends on the use case. (And I'm sorry I'm not familiar with the use cases.)
How about preparing another function that takes for example a threshold for `confidence`, and returns multiple predicted paths?

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @kenji-miyake @yukkysaito So in this function, we return a single predicted path even when all predicted paths have the same confidence right? If so, I think we can return the predicted path that is in the front of the predicted paths

Comment (yukkysaito) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @purewater0901 Who is this function intended to be used by? planning?

Comment (mitsudome-r) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: I'm also concerned if it is really okay to return only the path with the best confidence.
e.g., If there are two candidate paths with 51% confidence and 49% confidence, is it really reasonable to only consider the path with 51%? 

My suggestion would be to take threshold as an argument for confidence like @kenji-miyake suggested, or  add argument `N` and let the function to return top N paths with descending confidence. 

Anyway, I think we should clarify why you would like to add this function. If you have any example use cases in your mind, could you provide them?

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @mitsudome-r @yukkysaito Yes, some planning modules use this function to pick up one predicted path(They need only one predicted path).

For example, obstacle cruise planner and lane change planner only select one predicted path. You can find their usage in the code. Please go to the these modules and take a look.

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: @mitsudome-r @yukkysaito 

example1: https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/src/scene_module/lane_change/util.cpp#L398-L403

example2: https://github.com/autowarefoundation/autoware.universe/blob/main/planning/obstacle_cruise_planner/src/node.cpp#L747

Comment (purewater0901) [reply common/perception_utils/test/src/test_predicted_path_utils.cpp:92]: These are examples that we need to select the best-predicted path for computing lane change path or vehicle velocity

**Comment by mitsudome-r** [comment]
I talked with @purewater0901 and decided to close this PR

**Pull request closed without merging**